### ISSUE TITLE
Bug: NaN result with projection to 1D latent space

### ISSUE THREAD

**Issue by rmitsch** [lmcinnes/umap#85]
Hi,

thanks for publishing and maintaining this cool project! I played around with UMAP a bit and encountered an oddity: I have a small (n=178) distance matrix that I want to project onto a 1D space. If I run UMAP in version 0.3.0 with the default parameters, sometimes the resulting low-dimensional data consists of `NaN`s exclusively while sometimes the result seems perfectly fine.

I narrowed it down to the `random_state ` being used, which seems a bit fishy. To reproduce: I call
```
res = umap.UMAP(
    n_components=1,
    random_state=8, 
    metric='precomputed'
).fit_transform(high_dim_data)

print(numpy.count_nonzero(numpy.isnan(res)))
```
where `high_dim_data` is as defined in https://gist.github.com/rmitsch/cd99fcacf6a8f3c0d41af6f903735251.

I didn't observe this behaviour with `n_components` > 1, but haven't checked thoroughly yet. I just tried various values for `random_state` until I could reproduce the `NaN` result.

Is this a known problem? I couldn't find any related issues.

**Comment by lmcinnes** [comment]
It is not a known problem. My best guess is that the spectral
initialization is going badly for your data. As a workaround you can add
``init='random'`` to your UMAP parameters and that may fix the issue for
now. I'll have to try and track down what is actually going wrong here.
Thanks for the data and a random seed -- that will help a great deal with
reproducing the problem and hopefully fixing it!

**Comment by rmitsch** [comment]
Thanks for taking a look into it.

**Comment by asanakoy** [comment]
I faced the same issue when trying to run UMAP with `n_components=1` with precomputed metric.
It works from time to time, but in most of the cases it prodices NaN embeddings.   
Dataset size 1000x1000.    
Setting `init='random` didn't help.

UMAP was installed from git commit 91431b29049f74d66b575d1e632576a6642b9213

**Comment by lmcinnes** [comment]
Thanks for the report -- that sounds like something that is more likely to
be reproducible. I'm on vacation right now but I'll try to look into this
when I get back.

**Comment by lmcinnes** [comment]
So the bad news is that I grabbed the data from the gist and ran with the supplied parameters but can't reproduce the error locally here. That, unfortunately, makes it hard to track down the actual issue. I'll keep a lookout, but I can't make any promises on getting this fixed too soon as it seems to be intermittent/hard to reproduce.

**Comment by asanakoy** [comment]
@lmcinnes have you tried to run several times with different random initializations? It is non-deterministic. Sometimes works, sometimes not.

**Comment by lmcinnes** [comment]
Fine suggestion. I set a loop running until I got a failure, which is now reproducible with that seed. I'll try to dig into this soon and see what the actual underlying issues is. Thanks!

**Comment by snakers4** [comment]
> I faced the same issue when trying to run UMAP with n_components=1 with precomputed metric.
> It works from time to time, but in most of the cases it prodices NaN embeddings.
> Dataset size 1000x1000.
> Setting init='random didn't help.

Faced a similar issue - but then doing something like 300k * 50 => 300k * 2
When doing 300k * 50 => 300k * 10 faced no errors whatsoever

**Comment by lmcinnes** [comment]
Got it! Well, at lest the one that I could reliably reproduce. If distinct points ended up having perfectly identical coordinates during the embedding optimization that could trigger problems. I believe I now have a catch in place for that. I would be happy if any of the reporters (especially with reproducible issues) could verify that the current master now resolves the problem for them.

**Comment by rmitsch** [comment]
Fixes my problem. The version currently available on pip (0.1.1) does not contain the fix though, does it? I just copied the changes in your commit by hand for now.

Thanks!

**Comment by asanakoy** [comment]
Fixed the problem for me. Installed from master 16e7ce624dc984cb4a626d7cdfccff4bd8e7ce2c
